File: CFNDeploy/java/src/main/java/com/example/AppStack.java
package com.example;

/**
 * Application infrastructure stack.
 */

import software.amazon.awscdk.CfnOutput;
import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.services.autoscaling.AutoScalingGroup;
import software.amazon.awscdk.services.codedeploy.ServerApplication;
import software.amazon.awscdk.services.codedeploy.ServerDeploymentGroup;
import software.amazon.awscdk.services.codedeploy.InstanceTagSet;
import software.amazon.awscdk.services.ec2.*;
import software.amazon.awscdk.services.iam.ManagedPolicy;
import software.amazon.awscdk.services.iam.Role;
import software.amazon.awscdk.services.iam.ServicePrincipal;
import software.constructs.Construct;
import java.util.Arrays;
import java.util.Collections;
import java.util.Map;

public class AppStack extends Stack {
    public AppStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

        Vpc vpc = Vpc.Builder.create(this, "AppVPC").maxAzs(2).natGateways(1).build();
        
        SecurityGroup securityGroup = SecurityGroup.Builder.create(this, "AppSecurityGroup")
                .vpc(vpc).description("Security group for application instances").allowAllOutbound(true).build();
        
        securityGroup.addIngressRule(Peer.anyIpv4(), Port.tcp(80), "Allow HTTP");
        securityGroup.addIngressRule(Peer.anyIpv4(), Port.tcp(22), "Allow SSH");
        
        Role instanceRole = Role.Builder.create(this, "InstanceRole")
                .assumedBy(new ServicePrincipal("ec2.amazonaws.com"))
                .managedPolicies(Arrays.asList(
                        ManagedPolicy.fromAwsManagedPolicyName("CloudWatchAgentServerPolicy"),
                        ManagedPolicy.fromAwsManagedPolicyName("AmazonSSMManagedInstanceCore")))
                .build();
        
        LaunchTemplate launchTemplate = LaunchTemplate.Builder.create(this, "AppLaunchTemplate")
                .instanceType(InstanceType.of(InstanceClass.T3, InstanceSize.MICRO))
                .machineImage(MachineImage.latestAmazonLinux2())
                .securityGroup(securityGroup).role(instanceRole).userData(UserData.forLinux()).build();
        
        launchTemplate.getUserData().addCommands(
                "yum update -y", "yum install -y httpd ruby wget", "systemctl start httpd",
                "systemctl enable httpd", "echo '<h1>Hello from CDK Pipeline!</h1>' > /var/www/html/index.html",
                "cd /home/ec2-user", "wget https://aws-codedeploy-us-east-1.s3.us-east-1.amazonaws.com/latest/install",
                "chmod +x ./install", "./install auto");
        
        AutoScalingGroup asg = AutoScalingGroup.Builder.create(this, "AppASG")
                .vpc(vpc).launchTemplate(launchTemplate).minCapacity(1).maxCapacity(3).desiredCapacity(2)
                .vpcSubnets(SubnetSelection.builder().subnetType(SubnetType.PRIVATE_WITH_EGRESS).build()).build();
        
        ServerApplication application = ServerApplication.Builder.create(this, "AppApplication")
                .applicationName("my-app-" + this.getStackName()).build();
        
        ServerDeploymentGroup deploymentGroup = ServerDeploymentGroup.Builder.create(this, "AppDeploymentGroup")
                .application(application).deploymentGroupName("production").autoScalingGroups(Collections.singletonList(asg))
                .installAgent(true).ec2InstanceTags(new InstanceTagSet(Map.of("Environment", Arrays.asList("production")))).build();
        
        CfnOutput.Builder.create(this, "VPCId").value(vpc.getVpcId()).description("VPC ID").build();
        CfnOutput.Builder.create(this, "ApplicationName").value(application.getApplicationName()).description("CodeDeploy Application Name").build();
        CfnOutput.Builder.create(this, "DeploymentGroupName").value(deploymentGroup.getDeploymentGroupName()).description("CodeDeploy Deployment Group Name").build();
    }
}

File: ECRBuild/csharp/src/AppStack.cs
using Amazon.CDK;
using Amazon.CDK.AWS.EC2;
using Amazon.CDK.AWS.AutoScaling;
using Amazon.CDK.AWS.CodeDeploy;
using Amazon.CDK.AWS.IAM;
using Constructs;

/// <summary>
/// Application infrastructure stack.
/// </summary>
namespace ECRBuild
{
    public class AppStack : Stack
    {
        internal AppStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
        {
            var vpc = new Vpc(this, "AppVPC", new VpcProps { MaxAzs = 2, NatGateways = 1 });
            
            var securityGroup = new SecurityGroup(this, "AppSecurityGroup", new SecurityGroupProps
            {
                Vpc = vpc, Description = "Security group for application instances", AllowAllOutbound = true
            });
            
            securityGroup.AddIngressRule(Peer.AnyIpv4(), Port.Tcp(80), "Allow HTTP");
            securityGroup.AddIngressRule(Peer.AnyIpv4(), Port.Tcp(22), "Allow SSH");
            
            var instanceRole = new Role(this, "InstanceRole", new RoleProps
            {
                AssumedBy = new ServicePrincipal("ec2.amazonaws.com"),
                ManagedPolicies = new[] {
                    ManagedPolicy.FromAwsManagedPolicyName("CloudWatchAgentServerPolicy"),
                    ManagedPolicy.FromAwsManagedPolicyName("AmazonSSMManagedInstanceCore")
                }
            });
            
            var launchTemplate = new LaunchTemplate(this, "AppLaunchTemplate", new LaunchTemplateProps
            {
                InstanceType = InstanceType.Of(InstanceClass.T3, InstanceSize.MICRO),
                MachineImage = MachineImage.LatestAmazonLinux2(),
                SecurityGroup = securityGroup, Role = instanceRole, UserData = UserData.ForLinux()
            });
            
            launchTemplate.UserData.AddCommands(
                "yum update -y", "yum install -y httpd ruby wget", "systemctl start httpd",
                "systemctl enable httpd", "echo '<h1>Hello from CDK Pipeline!</h1>' > /var/www/html/index.html",
                "cd /home/ec2-user", "wget https://aws-codedeploy-us-east-1.s3.us-east-1.amazonaws.com/latest/install",
                "chmod +x ./install", "./install auto");
            
            var asg = new AutoScalingGroup(this, "AppASG", new AutoScalingGroupProps
            {
                Vpc = vpc, LaunchTemplate = launchTemplate, MinCapacity = 1, MaxCapacity = 3, DesiredCapacity = 2,
                VpcSubnets = new SubnetSelection { SubnetType = SubnetType.PRIVATE_WITH_EGRESS }
            });
            
            var application = new ServerApplication(this, "AppApplication", new ServerApplicationProps
            {
                ApplicationName = $"my-app-{this.StackName}"
            });
            
            var deploymentGroup = new ServerDeploymentGroup(this, "AppDeploymentGroup", new ServerDeploymentGroupProps
            {
                Application = application, DeploymentGroupName = "production", AutoScalingGroups = new[] { asg },
                InstallAgent = true, Ec2InstanceTags = new InstanceTagSet(new System.Collections.Generic.Dictionary<string, string[]>
                {
                    { "Environment", new[] { "production" } }
                })
            });
            
            new CfnOutput(this, "VPCId", new CfnOutputProps { Value = vpc.VpcId, Description = "VPC ID" });
            new CfnOutput(this, "ApplicationName", new CfnOutputProps { Value = application.ApplicationName, Description = "CodeDeploy Application Name" });
            new CfnOutput(this, "DeploymentGroupName", new CfnOutputProps { Value = deploymentGroup.DeploymentGroupName, Description = "CodeDeploy Deployment Group Name" });
        }
    }
}
